motionplan/rrtStarConnect.go

```diff
@@ -176,64 +225,138 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
 		}
 
 		// get next sample, switch map pointers
-		target = referenceframe.RandomFrameInputs(mp.frame, mp.randseed)
+		target, err = mp.sample(map1reached, i)
+		if err != nil {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+			return
+		}
+		map1, map2 = map2, map1
 	}
 	mp.logger.Debug("RRT* exceeded max iter")
 	rrt.solutionChan <- shortestPath(rrt.maps, shared)
 }
 
-func (mp *rrtStarConnectMotionPlanner) extend(
-	tree rrtMap,
-	target []referenceframe.Input,
+func (mp *rrtStarConnectMotionPlanner) sample(rSeed node, sampleNum int) ([]referenceframe.Input, error) {
+	// If we have done more than 50 iterations, start seeding off completely random positions 2 at a time
+	// The 2 at a time is to ensure random seeds are added onto both the seed and goal maps.
+	if rSeed == nil || (sampleNum >= mp.algOpts.IterBeforeRand && sampleNum%4 >= 2) {
+		return referenceframe.RandomFrameInputs(mp.frame, mp.randseed), nil
+	}
+	// Seeding nearby to valid points results in much faster convergence in less constrained space
+	return referenceframe.RestrictedRandomFrameInputs(mp.frame, mp.randseed, 0.1, rSeed.Q())
+}
+
+func (mp *rrtStarConnectMotionPlanner) constrainedExtend(
+	ctx context.Context,
+	rrtMap map[node]node,
+	near, target node,
 	mchan chan node,
 ) {
-	if validTarget := mp.checkInputs(target); !validTarget {
-		mchan <- nil
-		return
-	}
-	// iterate over the k nearest neighbors and find the minimum cost to connect the target node to the tree
-	neighbors := kNearestNeighbors(mp.planOpts, tree, &basicNode{q: target}, mp.algOpts.NeighborhoodSize)
```

Comment:
I'm not convinced this is going to rewire correctly if you aren't using kNearestNeighbors